# RLlib Sample Application

In [20]:
# this first code cell is a little magic to install Ray 
# properly, due to hardcoded limitations in TensorFlow

!pip uninstall -y -q pyarrow
!pip install -q ray[debug]
!pip install ray[tune]
!pip install lz4

try:
  %tensorflow_version 2.x
except Exception:
  pass

Then we start Ray…

In [9]:
import ray
import ray.rllib.agents.ppo as ppo

ray.shutdown()
ray.init(ignore_reinit_error=True)

2020-03-05 16:26:05,159	INFO resource_spec.py:212 -- Starting Ray with 3.47 GiB memory available for workers and up to 1.74 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-03-05 16:26:05,448	INFO services.py:1078 -- View the Ray dashboard at localhost:8265


{'node_ip_address': '192.168.1.65',
 'redis_address': '192.168.1.65:35051',
 'object_store_address': '/tmp/ray/session_2020-03-05_16-26-05_143012_68388/sockets/plasma_store',
 'raylet_socket_name': '/tmp/ray/session_2020-03-05_16-26-05_143012_68388/sockets/raylet',
 'webui_url': 'localhost:8265',
 'session_dir': '/tmp/ray/session_2020-03-05_16-26-05_143012_68388'}

After a successful launch, there should be a log output line that reads something to the effect of `View the Ray dashboard at localhost:8265` in which case open another browser tab for the Ray dashboard at <http://localhost:8265/>

Next we'll train an RLlib policy with the `CartPole-v0` environment, which is a relatively simple and quick example. For more details about this problem, see the tutorial [*Cartpole - Introduction to Reinforcement Learning (DQN - Deep Q-Learning)*](https://towardsdatascience.com/cartpole-introduction-to-reinforcement-learning-ed0eb5b58288) by [Greg Surma](https://twitter.com/GSurma).

In [22]:
config = ppo.DEFAULT_CONFIG.copy()
config["log_level"] = "WARN"

n_iter = 10
reward_history = []

agent = ppo.PPOTrainer(config, env="CartPole-v0")

for _ in range(n_iter):
    result = agent.train()
    print(result)

    max_reward = result["episode_reward_max"]
    reward_history.append(max_reward)

    file_name = agent.save("/tmp/ppo")
    print(f"\n{file_name}")

/opt/anaconda3/lib/python3.7/site-packages/ray/rllib/utils/from_config.py:134: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  obj = yaml.load(type_)
2020-03-05 16:52:57,522	WARNING util.py:37 -- Install gputil for GPU system monitoring.


(pid=68816) /opt/anaconda3/lib/python3.7/site-packages/ray/rllib/utils/from_config.py:134: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
(pid=68816)   obj = yaml.load(type_)
(pid=68815) /opt/anaconda3/lib/python3.7/site-packages/ray/rllib/utils/from_config.py:134: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
(pid=68815)   obj = yaml.load(type_)
{'episode_reward_max': 71.0, 'episode_reward_min': 9.0, 'episode_reward_mean': 22.491525423728813, 'episode_len_mean': 22.491525423728813, 'episodes_this_iter': 177, 'policy_reward_min': {}, 'policy_reward_max': {}, 'policy_reward_mean': {}, 'custom_metrics': {}, 'hist_stats': {'episode_reward': [29.0, 15.0, 18.0, 13.0, 42.0, 15.0, 16.0, 22.0, 13.0, 18.0, 12.0, 23.0, 14.0, 17.0, 35.0, 14.0, 12.0, 14.0, 26.0, 27.0, 46

{'episode_reward_max': 200.0, 'episode_reward_min': 12.0, 'episode_reward_mean': 64.75, 'episode_len_mean': 64.75, 'episodes_this_iter': 41, 'policy_reward_min': {}, 'policy_reward_max': {}, 'policy_reward_mean': {}, 'custom_metrics': {}, 'hist_stats': {'episode_reward': [141.0, 156.0, 166.0, 69.0, 142.0, 149.0, 106.0, 31.0, 59.0, 52.0, 55.0, 64.0, 109.0, 112.0, 110.0, 200.0, 44.0, 154.0, 58.0, 41.0, 106.0, 130.0, 12.0, 135.0, 76.0, 113.0, 66.0, 39.0, 59.0, 89.0, 71.0, 200.0, 19.0, 94.0, 194.0, 31.0, 200.0, 106.0, 20.0, 103.0, 27.0, 46.0, 46.0, 60.0, 19.0, 13.0, 34.0, 28.0, 23.0, 37.0, 25.0, 86.0, 14.0, 50.0, 18.0, 34.0, 19.0, 13.0, 17.0, 59.0, 43.0, 32.0, 14.0, 36.0, 49.0, 56.0, 46.0, 22.0, 45.0, 31.0, 82.0, 27.0, 95.0, 47.0, 61.0, 46.0, 24.0, 67.0, 75.0, 41.0, 13.0, 62.0, 16.0, 46.0, 44.0, 14.0, 60.0, 75.0, 31.0, 153.0, 21.0, 15.0, 46.0, 19.0, 92.0, 16.0, 22.0, 76.0, 117.0, 49.0], 'episode_lengths': [141, 156, 166, 69, 142, 149, 106, 31, 59, 52, 55, 64, 109, 112, 110, 200, 44, 154, 5

{'episode_reward_max': 200.0, 'episode_reward_min': 12.0, 'episode_reward_mean': 125.95, 'episode_len_mean': 125.95, 'episodes_this_iter': 23, 'policy_reward_min': {}, 'policy_reward_max': {}, 'policy_reward_mean': {}, 'custom_metrics': {}, 'hist_stats': {'episode_reward': [151.0, 200.0, 200.0, 200.0, 200.0, 200.0, 67.0, 200.0, 83.0, 200.0, 154.0, 67.0, 200.0, 200.0, 170.0, 200.0, 200.0, 200.0, 111.0, 200.0, 177.0, 127.0, 200.0, 75.0, 31.0, 153.0, 21.0, 15.0, 46.0, 19.0, 92.0, 16.0, 22.0, 76.0, 117.0, 49.0, 141.0, 156.0, 166.0, 69.0, 142.0, 149.0, 106.0, 31.0, 59.0, 52.0, 55.0, 64.0, 109.0, 112.0, 110.0, 200.0, 44.0, 154.0, 58.0, 41.0, 106.0, 130.0, 12.0, 135.0, 76.0, 113.0, 66.0, 39.0, 59.0, 89.0, 71.0, 200.0, 19.0, 94.0, 194.0, 31.0, 200.0, 106.0, 20.0, 103.0, 27.0, 200.0, 200.0, 200.0, 200.0, 169.0, 180.0, 166.0, 200.0, 161.0, 162.0, 162.0, 137.0, 200.0, 200.0, 179.0, 200.0, 86.0, 99.0, 200.0, 200.0, 148.0, 199.0, 200.0], 'episode_lengths': [151, 200, 200, 200, 200, 200, 67, 200, 83

{'episode_reward_max': 200.0, 'episode_reward_min': 19.0, 'episode_reward_mean': 171.65, 'episode_len_mean': 171.65, 'episodes_this_iter': 21, 'policy_reward_min': {}, 'policy_reward_max': {}, 'policy_reward_mean': {}, 'custom_metrics': {}, 'hist_stats': {'episode_reward': [200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 125.0, 119.0, 200.0, 200.0, 200.0, 200.0, 59.0, 89.0, 71.0, 200.0, 19.0, 94.0, 194.0, 31.0, 200.0, 106.0, 20.0, 103.0, 27.0, 200.0, 200.0, 200.0, 200.0, 169.0, 180.0, 166.0, 200.0, 161.0, 162.0, 162.0, 137.0, 200.0, 200.0, 179.0, 200.0, 86.0, 99.0, 200.0, 200.0, 148.0, 199.0, 200.0, 151.0, 200.0, 200.0, 200.0, 200.0, 200.0, 67.0, 200.0, 83.0, 200.0, 154.0, 67.0, 200.0, 200.0, 170.0, 200.0, 200.0, 200.0, 111.0, 200.0, 177.0, 127.0, 200.0, 200.0, 200.0, 200.0, 174.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 194.0, 185.0, 200.0], 'episode_lengths': [200, 200, 200, 20

{'episode_reward_max': 200.0, 'episode_reward_min': 67.0, 'episode_reward_mean': 190.94, 'episode_len_mean': 190.94, 'episodes_this_iter': 20, 'policy_reward_min': {}, 'policy_reward_max': {}, 'policy_reward_mean': {}, 'custom_metrics': {}, 'hist_stats': {'episode_reward': [200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 67.0, 200.0, 83.0, 200.0, 154.0, 67.0, 200.0, 200.0, 170.0, 200.0, 200.0, 200.0, 111.0, 200.0, 177.0, 127.0, 200.0, 200.0, 200.0, 200.0, 174.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 194.0, 185.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 125.0, 119.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 141.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0], 'episode_lengths': [200, 20

In [23]:
print(reward_history)

[71.0, 153.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0, 200.0]


The history of `max_reward` shows that this model `200` by the third iteration -- which is good, since the [*solution*](https://gym.openai.com/envs/CartPole-v0/) for `CartPole-v0` is to get an average reward of `195.0` over a hundred consecutive trials.

In [24]:
! rllib rollout \
    /tmp/ppo/checkpoint_10/checkpoint-10 \
    --config "{\"env\": \"CartPole-v0\"}" --run PPO \
    --steps 2000

2020-03-05 16:54:18,677	INFO resource_spec.py:212 -- Starting Ray with 3.37 GiB memory available for workers and up to 1.71 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-03-05 16:54:19,063	INFO services.py:1078 -- View the Ray dashboard at localhost:8266
2020-03-05 16:54:19,663	INFO trainer.py:420 -- Tip: set 'eager': true or the --eager flag to enable TensorFlow eager execution
2020-03-05 16:54:19,705	INFO trainer.py:580 -- Current log_level is WARN. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
/opt/anaconda3/lib/python3.7/site-packages/ray/rllib/utils/from_config.py:134: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  obj = yaml.load(type_)
2020-03-05 16:54:24,252	WARNING util.py:37 -- Install gputil for GPU system monitoring.
2020-03-05 16:54:24,344	WARNING trainab

Now that we've trained a model, we can look at its resulting policy…

In [25]:
import pprint

policy = agent.get_policy()
model = policy.model

pprint.pprint(model.variables())
pprint.pprint(model.value_function())

[<tf.Variable 'default_policy/fc_1/kernel:0' shape=(4, 256) dtype=float32>,
 <tf.Variable 'default_policy/fc_1/bias:0' shape=(256,) dtype=float32>,
 <tf.Variable 'default_policy/fc_value_1/kernel:0' shape=(4, 256) dtype=float32>,
 <tf.Variable 'default_policy/fc_value_1/bias:0' shape=(256,) dtype=float32>,
 <tf.Variable 'default_policy/fc_2/kernel:0' shape=(256, 256) dtype=float32>,
 <tf.Variable 'default_policy/fc_2/bias:0' shape=(256,) dtype=float32>,
 <tf.Variable 'default_policy/fc_value_2/kernel:0' shape=(256, 256) dtype=float32>,
 <tf.Variable 'default_policy/fc_value_2/bias:0' shape=(256,) dtype=float32>,
 <tf.Variable 'default_policy/fc_out/kernel:0' shape=(256, 2) dtype=float32>,
 <tf.Variable 'default_policy/fc_out/bias:0' shape=(2,) dtype=float32>,
 <tf.Variable 'default_policy/value_out/kernel:0' shape=(256, 1) dtype=float32>,
 <tf.Variable 'default_policy/value_out/bias:0' shape=(1,) dtype=float32>]
<tf.Tensor 'Reshape:0' shape=(?,) dtype=float32>


In [26]:
model.base_model.summary()

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
observations (InputLayer)       [(None, 4)]          0                                            
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 256)          1280        observations[0][0]               
__________________________________________________________________________________________________
fc_value_1 (Dense)              (None, 256)          1280        observations[0][0]               
__________________________________________________________________________________________________
fc_2 (Dense)                    (None, 256)          65792       fc_1[0][0]                       
______________________________________________________________________________________________